### PR TITLE
sdk/oak: Add Handle (un)pack for safe serialization to Wasm host ABI

### PR DESCRIPTION
**Pull request by blaxill**
- ~~Also adds missing resize for HANDLE_SPACE_TOO_SMALL~~

#451

### CODE REVIEW

**Comment by blaxill** [comment]
Looks like the HANDLE_SPACE_TOO_SMALL part of this will be covered by #453

File: sdk/rust/oak/src/lib.rs [hunk 1]
```diff
@@ -140,6 +163,10 @@ pub fn wait_on_channels(handles: &[ReadHandle]) -> Result<Vec<ChannelReadStatus>
 pub fn channel_read(half: ReadHandle, buf: &mut Vec<u8>, handles: &mut Vec<Handle>) -> OakStatus {
     // Try reading from the channel twice: first with provided vectors, then
     // with vectors that have been resized to meet size requirements.
+
+    // We cannot deserialize directly into the handles vector if `Handle` may
```
Comment (daviddrysdale) [inline sdk/rust/oak/src/lib.rs]: nit: maybe s/if/because/ ?

File: sdk/rust/oak/src/lib.rs [hunk 2]
```diff
@@ -140,6 +163,10 @@ pub fn wait_on_channels(handles: &[ReadHandle]) -> Result<Vec<ChannelReadStatus>
...
+    // not have the correct memory layout
```
Comment (daviddrysdale) [inline sdk/rust/oak/src/lib.rs]: super-nit: full stop at end of comment

File: sdk/rust/oak/src/lib.rs [hunk 3]
```diff
@@ -149,34 +176,69 @@ pub fn channel_read(half: ReadHandle, buf: &mut Vec<u8>, handles: &mut Vec<Handl
...
+
+                        // actual_handle_count is number of handles not bytes
+                        handles_buf.set_len(actual_handle_count as usize * 8);
+                        Handle::unpack(&handles_buf, actual_handle_count, handles);
                     };
                     return s;
                 }
-                OakStatus::ERR_BUFFER_TOO_SMALL if !(*resized) => {
+
+                // ERR_BUFFER_TOO_SMALL can shadow ERR_HANDLE_SPACE_TOO_SMALL
```
Comment (daviddrysdale) [inline sdk/rust/oak/src/lib.rs]: nit: comment leaves me thinking "so what?"

File: sdk/rust/oak/src/lib.rs [hunk 4]
```diff
@@ -149,34 +176,69 @@ pub fn channel_read(half: ReadHandle, buf: &mut Vec<u8>, handles: &mut Vec<Handl
...
+                OakStatus::ERR_BUFFER_TOO_SMALL | OakStatus::ERR_HANDLE_SPACE_TOO_SMALL
+                    if !(*resized) =>
+                {
```
Comment (daviddrysdale) [inline sdk/rust/oak/src/lib.rs]: We should sync up on whether to pull in #453 first or just use this &ndash; thoughts?

File: sdk/rust/oak/src/lib.rs [hunk 5]
```diff
@@ -61,6 +62,28 @@ impl Handle {
             id: oak_abi::INVALID_HANDLE,
         }
     }
+
+    /// Pack a slice of `Handles` into the Wasm host ABI format.
+    fn pack(handles: &[Handle]) -> Vec<u8> {
+        let mut packed = vec![];
```
[the remark this answers is not in the document]
Comment (blaxill) [reply sdk/rust/oak/src/lib.rs]: Done

File: sdk/rust/oak/src/lib.rs (hunk 3 above)
Comment (blaxill) [reply sdk/rust/oak/src/lib.rs]: For some clarification why the cases weren't split but I don't think it was important. I'm keeping your case logic from #453 close as possible

File: sdk/rust/oak/src/lib.rs (hunk 4 above)
Comment (blaxill) [reply sdk/rust/oak/src/lib.rs]: Rebasing on  #453

File: sdk/rust/oak/src/lib.rs (hunk 2 above)
Comment (blaxill) [reply sdk/rust/oak/src/lib.rs]: Done

File: sdk/rust/oak/src/lib.rs [hunk 6]
```diff
@@ -139,6 +162,10 @@ pub fn wait_on_channels(handles: &[ReadHandle]) -> Result<Vec<ChannelReadStatus>
 pub fn channel_read(half: ReadHandle, buf: &mut Vec<u8>, handles: &mut Vec<Handle>) -> OakStatus {
     // Try reading from the channel twice: first with provided vectors, then
     // with vectors that have been resized to meet size requirements.
+
+    // We cannot deserialize directly into the handles vector because `Handle` may
+    // not have the correct memory layout.
+    let mut handles_buf = vec![];
```
Comment (daviddrysdale) [inline sdk/rust/oak/src/lib.rs]: Should this be `Vec::with_capacity(handles.capacity() * 8)`, as the caller may well have accurate expectations about how many handles to expect?

File: sdk/rust/oak/src/lib.rs [hunk 7]
```diff
@@ -176,14 +207,15 @@ pub fn channel_read(half: ReadHandle, buf: &mut Vec<u8>, handles: &mut Vec<Handl
                         let extra = (actual_size as usize) - buf.len();
                         buf.reserve(extra);
                     }
+
+                    let handles_capacity = handles_buf.len() / 8;
```
Comment (daviddrysdale) [inline sdk/rust/oak/src/lib.rs]: should that be `.capacity() / 8` ?

File: sdk/rust/oak/src/lib.rs (hunk 6 above)
Comment (blaxill) [reply sdk/rust/oak/src/lib.rs]: Done

File: sdk/rust/oak/src/lib.rs (hunk 7 above)
Comment (blaxill) [reply sdk/rust/oak/src/lib.rs]: Done

**Pull request merged**